File: dot-net/hcx-integrator-sdk/Io.HcxProtocol/Impl/HCXOutgoingRequest.cs
using Io.HcxProtocol.Dto;
using Io.HcxProtocol.Exceptions;
using Io.HcxProtocol.Helper;
using Io.HcxProtocol.Init;
using Io.HcxProtocol.Interfaces;
using Io.HcxProtocol.Jwe;
using Io.HcxProtocol.Key;
using Io.HcxProtocol.Service;
using Io.HcxProtocol.Utils;
using NLog;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Security.Cryptography;
using System.IO;



namespace Io.HcxProtocol.Impl
{
    /**
     * Library  : Io.Hcx.Protocol
     * Author   : WalkingTree Technologies
     * Date     : 15-Mar-2023
     * All Rights Reserved. WalkingTree Technologies.
     **/

    /// <summary>
    ///     The <b>HCX Outgoing Request</b> class provide the methods to help in creating the JWE Payload and send the request to the sender system from HCX Gateway.
    /// </summary>
    /// <remarks>
    ///     <list type="bullet">
    ///         <item><b>Generate</b> is the main method to use by the integrator(s) to generate JWE Payload.
    ///         <list type="bullet">
    ///             <item>This method handles two types of requests. There are two implementations of <b>Generate</b> method to handle action, on_action API calls.
    ///                 <list type="number">
    ///                     <item>Sending an initial request of HCX workflow action.</item>
    ///                     <item>Sending a response to the incoming HCX workflow action.
    ///                         <list type="bullet">
    ///                             <item>The input request JWE should be used as <i>actionJwe</i>.</item>
    ///                         </list>
    ///                     </item>
    ///                 </list>
    ///             </item>
    ///         </list>
    ///         </item>
    ///         <item>
    ///             <b>ValidatePayload, CreateHeader, EncryptPayload, InitializeHCXCall</b> methods are used by <b>Generate</b> method to compose the JWE payload generation functionality.
    ///             These methods are available for the integrator(s) to use them based on different scenario(s) or use cases.
    ///         </item>
    ///     </list>
    /// </remarks>
    public class HCXOutgoingRequest : FhirPayload, IOutgoingRequest, IDisposable
    {

        private static NLog.Logger _logger = NLog.LogManager.GetCurrentClassLogger();
        public HCXOutgoingRequest()
        {
            
        }
        //public virtual bool Process(string fhirPayload, Operations operation, string recipientCode, string apiCallId, string correlationId,string actionJwe, string onActionStatus, Dictionary<string, Object> domainHeaders, Dictionary<string, Object> output, Config config)
        //{
        //    bool result = false;
        //    try
        //    {
        //        Dictionary<string, Object> error = new Dictionary<string, Object>();
        //        Dictionary<string, Object> headers = new Dictionary<string, Object>(domainHeaders);
        //        Dictionary<string, Object> response = new Dictionary<string, Object>();

        //        if (config.FhirValidationEnabled)  //change ismail
        //        {
        //            if (!ValidatePayload(fhirPayload, operation, error, config))
        //            {
        //                foreach (var err in error) output.Add(err.Key, err.Value);
        //            }
        //        }
        //        if (!CreateHeader(config.ParticipantCode, recipientCode, apiCallId, correlationId, actionJwe, onActionStatus, headers, error))
        //        {
        //            foreach (var err in headers) output.Add(err.Key, err.Value);
        //        }
        //        else if (!EncryptPayload(headers, fhirPayload, output, config))
        //        {
        //            foreach (var err in error) output.Add(err.Key, err.Value);
        //        }
        //        else
        //        {
        //            result = InitializeHCXCall(JSONUtils.Serialize(output), operation, response, config);
        //            foreach (var err in response) output.Add(err.Key, err.Value);
        //        }
        //        return result;
        //    }
        //    catch (Exception ex)
        //    {
        //        // TODO: Exception is handled as domain processing error, we will be enhancing in next version.
        //        output.Add(ErrorCodes.ERR_DOMAIN_PROCESSING.ToString(), ex.Message.ToString());
        //        _logger.Error(ErrorCodes.ERR_DOMAIN_PROCESSING.ToString(), ex.Message.ToString());
        //        return result;
        //    }
        //}

        public virtual  bool Process(string fhirPayload, Operations operation, string recipientCode, string apiCallId, string correlationId, string actionJwe, string onActionStatus, Dictionary<string, Object> domainHeaders, Dictionary<string, Object> output, Config config)
        {
            return ProcessOutgoing(fhirPayload, operation, recipientCode, apiCallId, correlationId, "", actionJwe, onActionStatus, domainHeaders, output, config);
        }

        public  virtual bool Process(string fhirPayload, Operations operation, string recipientCode, string apiCallId, string correlationId, string workflowId, string actionJwe, string onActionStatus, Dictionary<string, Object> domainHeaders, Dictionary<string, Object> output, Config config)
        {
            return ProcessOutgoing(fhirPayload, operation, recipientCode, apiCallId, correlationId, workflowId, actionJwe, onActionStatus, domainHeaders, output, config);
        }


        public  bool ProcessOutgoing(string fhirPayload, Operations operation, string recipientCode, string apiCallId, string correlationId, string workflowId, string actionJwe, string onActionStatus, Dictionary<string, Object> domainHeaders, Dictionary<string, Object> output, Config config)
        {
            bool result = false;
            try
            {
                Dictionary<string, Object> error = new Dictionary<string, Object>();
                Dictionary<string, Object> response = new Dictionary<string, Object>();
                Dictionary<string, Object> headers = new Dictionary<string, Object>(domainHeaders);
               // _logger.Info("Processing outgoing request has started :: operation: {}", operation);

                if (config.FhirValidationEnabled)  //change ismail
                {
                    if (!ValidatePayload(fhirPayload, operation, error, config))
                    {
                        foreach (var err in error) output.Add(err.Key, err.Value);
                    }
                }
                else if (!CreateHeader(config.ParticipantCode, recipientCode, apiCallId, correlationId,workflowId, actionJwe, onActionStatus, headers, error))
                {
                    foreach (var err in headers) output.Add(err.Key, err.Value);
                }
                else if (!EncryptPayload(headers, fhirPayload, output, config))
                {
                    foreach (var err in error) output.Add(err.Key, err.Value);
                }
                else
                {
                    result = InitializeHCXCall(JSONUtils.Serialize(output), operation, response, config);
                    foreach (var res in response) output.Add(res.Key, res.Value);
                }
                if (output.ContainsKey(Constants.ERROR) || output.ContainsKey(ErrorCodes.ERR_DOMAIN_PROCESSING.ToString()))
                {
                    _logger.Error("Error while processing the outgoing request: {}", output);
                }
                return result;
            }
            catch (Exception ex)
            {
                // TODO: Exception is handled as domain processing error, we will be enhancing in next version.
                Console.WriteLine(ex.ToString());
                Console.Write(ex.StackTrace);
                output[ErrorCodes.ERR_DOMAIN_PROCESSING.ToString()] = ex.Message;
                _logger.Error("Error while processing the outgoing request: {}", ex.Message);
                return result;
            }
        }


        public virtual  bool ValidatePayload(string fhirPayload, Operations operation, Dictionary<string, Object> error, Config config)
        {
                return ValidateFHIR(fhirPayload, operation, error, config);                   
        }

        public virtual bool CreateHeader(string senderCode, string recipientCode, string apiCallId, string correlationId,string workflowId,  string actionJwe, string onActionStatus, Dictionary<string, Object> headers, Dictionary<string, Object> error)
        {
            try
            {
                //Note: [ALG, A256GCM] & [ENC, RSA_OAEP] headers added by JWE-EncryptRequest Method

                headers.Add(Constants.HCX_TIMESTAMP, DateTime.Now.ToString("yyyy-MM-dd'T'HH:mm:ss") + DateTime.Now.ToString("zzz").Replace(":", ""));

                if (string.IsNullOrEmpty(apiCallId))
                    apiCallId = Guid.NewGuid().ToString();
                headers.Add(Constants.HCX_API_CALL_ID, apiCallId);

                if (!string.IsNullOrEmpty(recipientCode))
                {
                    headers.Add(Constants.HCX_SENDER_CODE, senderCode);
                    headers.Add(Constants.HCX_RECIPIENT_CODE, recipientCode);
                    if (string.IsNullOrEmpty(correlationId))
                        correlationId = Guid.NewGuid().ToString();
                    headers.Add(Constants.HCX_CORRELATION_ID, correlationId);

                    if (!string.IsNullOrEmpty(workflowId))    //ismail CR No9
                        headers.Add(Constants.WORKFLOW_ID,workflowId);
                   
                }
                else
                {
                    Dictionary<string, Object> actionHeaders = JSONUtils.DecodeBase64String<Dictionary<string, Object>>(actionJwe.Split('.')[0]);
                    headers.Add(Constants.HCX_SENDER_CODE, actionHeaders[Constants.HCX_RECIPIENT_CODE]);
                    headers.Add(Constants.HCX_RECIPIENT_CODE, actionHeaders[Constants.HCX_SENDER_CODE]);
                    headers.Add(Constants.HCX_CORRELATION_ID, actionHeaders[Constants.HCX_CORRELATION_ID]);
                    headers.Add(Constants.STATUS, onActionStatus);
                    if (actionHeaders.ContainsKey(Constants.WORKFLOW_ID))
                        headers.Add(Constants.WORKFLOW_ID, actionHeaders[Constants.WORKFLOW_ID].ToString());
                }
                return true;
            }
            catch (Exception ex)
            {
                headers.Add(Constants.ERROR, "Error while creating headers: " + ex.Message.ToString());
                _logger.Error("Error while creating headers: " + ex.Message.ToString());
                return false;
            }
        }


        public virtual bool EncryptPayload(Dictionary<string, Object> headers, string fhirPayload, Dictionary<string, Object> encryptPayload, Config config)
        {
            try
            {
               
                //  string authToken = HcxUtils.GenerateToken(config.UserName, config.Password, config.AuthBasePath);
                string authToken = HcxUtils.GenerateToken(config);
                string publicKeyUrl = (string)HcxUtils.SearchRegistry((string)headers[Constants.HCX_RECIPIENT_CODE], authToken, config.ProtocolBasePath)[Constants.ENCRYPTION_CERT];

                RSA rsaPublicKey = X509KeyLoader.GetRSAPublicKeyFromPem(publicKeyUrl, PemMode.Url);
                JweRequest jweRequest = new JweRequest(headers, JSONUtils.Deserialize<Dictionary<string, Object>>(fhirPayload));
                jweRequest.EncryptRequest(rsaPublicKey);
                encryptPayload.Add(jweRequest.GetEncryptedObject().First().Key, jweRequest.GetEncryptedObject().First().Value);
                return true;
            }
            catch (Exception ex)
            {
                encryptPayload.Add(Constants.ERROR, ex.Message.ToString());
                _logger.Error("Error while encrypting the payload: {}" + ex.Message.ToString());
                return false;
            }
        }

        // we are handling the Exception in Process method
        public virtual bool InitializeHCXCall(string jwePayload, Operations operation, Dictionary<string, Object> response, Config config)
        {
            bool result = false;
            try
            {
                result = HcxUtils.initializeHCXCall(jwePayload, operation, response, config);
                //Dictionary<string, string> headers = new Dictionary<string, string>();
                //headers.Add(Constants.AUTHORIZATION, "Bearer " + HcxUtils.GenerateToken(config.UserName, config.Password, config.AuthBasePath));
                //HttpResponse hcxResponse = HttpUtils.Post(config.ProtocolBasePath + operation.getOperation(), headers, jwePayload);
                //response.Add(Constants.RESPONSE_OBJ, JSONUtils.Deserialize<Dictionary<string, Object>>(hcxResponse.Body));              

                //if (hcxResponse.Status == 202)
                //{
                //    result = true;
                //}
                //else
                //{
                //    _logger.Error("Error while processing the outgoing request::status: { } ::response: { }", hcxResponse.Status, response[Constants.RESPONSE_OBJ]);

                //}
            }
            catch(WebException ex)
            {
                _logger.Error(ex.Message);
            }
            return result;
        }

        private bool disposedValue;
        protected  void Dispose(bool disposing)
        {
            // check if already disposed
            if (!disposedValue)
            {
                if (disposing)
                {
                    // free managed Objects here
                }
                // free unmanaged Objects here
                // set the bool value to true
                disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(disposing: true);
            GC.SuppressFinalize(this);
        }

        public bool SendNotification(string topicCode, string recipientType, List<string> recipients, string message, Dictionary<string, string> templateParams, string correlationId, Dictionary<string, Object> output, Config config)
        {
            try
            {
                NotificationRequest notificationRequest = new NotificationRequest(topicCode, message, templateParams, recipientType, recipients, correlationId, config);
                NotificationService.validateNotificationRequest(notificationRequest);
                Dictionary<string, Object> requestBody = NotificationService.CreateNotificationRequest(notificationRequest, output, message);
                return InitializeHCXCall(JSONUtils.Serialize(requestBody), Operations.NOTIFICATION_NOTIFY, output, config);

            }
            catch (Exception e)
            {
                _logger.Error("Error while sending the notification: {}", e.Message);
                output.Add(Constants.ERROR, "Error while sending the notifications: " + e.Message);
                return false;
            }

        }

        ~HCXOutgoingRequest() { Dispose(disposing: false); }
    }
}

File: java/hcx-integrator-sdk/src/main/java/io/hcxprotocol/impl/HCXOutgoingRequest.java
package io.hcxprotocol.impl;

import com.typesafe.config.Config;
import io.hcxprotocol.dto.NotificationRequest;
import io.hcxprotocol.exception.ErrorCodes;
import io.hcxprotocol.helper.FhirPayload;
import io.hcxprotocol.interfaces.OutgoingRequest;
import io.hcxprotocol.jwe.JweRequest;
import io.hcxprotocol.key.PublicKeyLoader;
import io.hcxprotocol.service.NotificationService;
import io.hcxprotocol.utils.Constants;
import io.hcxprotocol.utils.JSONUtils;
import io.hcxprotocol.utils.Operations;
import io.hcxprotocol.utils.Utils;
import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.interfaces.RSAPublicKey;
import java.text.SimpleDateFormat;
import java.util.*;

/**
 * The <b>HCX Outgoing Request</b> class provide the methods to help in creating the JWE Payload and send the request to the sender system from HCX Gateway.
 *
 * <ul>
 *     <li><b>generate</b> is the main method to use by the integrator(s) to generate JWE Payload.
 *     <ul>
 *         <li>This method handles two types of requests. There are two implementations of <b>generate</b> method to handle action, on_action API calls.
 *         <ol>
 *            <li>
 *              Sending an initial request of HCX workflow action.
 *            </li>
 *            <li>Sending a response to the incoming HCX workflow action.
 *              <ul>
 *                  <li>The input request JWE should be used as <i>actionJwe</i>.</li>
 *              </ul>
 *            </li>
 *        </ol></li>
 *     </ul></li>
 *     <li>
 *         <b>validatePayload, createHeader, encryptPayload, initializeHCXCall</b> methods are used by <b>generate</b> method to compose the JWE payload generation functionality.
 *         These methods are available for the integrator(s) to use them based on different scenario(s) or use cases.
 *     </li>
 * </ul>
 *
 */
public class HCXOutgoingRequest extends FhirPayload implements OutgoingRequest {

    private static final Logger logger = LoggerFactory.getLogger(HCXOutgoingRequest.class);

    private NotificationService notificationService;

    @Override
    public boolean process(String fhirPayload, Operations operation, String recipientCode, String apiCallId, String correlationId, String actionJwe, String onActionStatus, Map<String, Object> domainHeaders, Map<String, Object> output, Config config) {
        return processOutgoing(fhirPayload, operation, recipientCode, apiCallId, correlationId, "", actionJwe, onActionStatus, domainHeaders, output, config);
    }

    @Override
    public boolean process(String fhirPayload, Operations operation, String recipientCode, String apiCallId, String correlationId, String workflowId, String actionJwe, String onActionStatus, Map<String, Object> domainHeaders, Map<String, Object> output, Config config) {
        return processOutgoing(fhirPayload, operation, recipientCode, apiCallId, correlationId, workflowId, actionJwe, onActionStatus, domainHeaders, output, config);
    }

    public boolean processOutgoing(String fhirPayload, Operations operation, String recipientCode, String apiCallId, String correlationId, String workflowId, String actionJwe, String onActionStatus, Map<String, Object> domainHeaders, Map<String, Object> output, Config config) {
        boolean result = false;
        try {
            Map<String, Object> error = new HashMap<>();
            Map<String, Object> response = new HashMap<>();
            Map<String, Object> headers = new HashMap<>(domainHeaders);
            logger.info("Processing outgoing request has started :: operation: {}", operation);
            if (!validatePayload(fhirPayload, operation, error, config)) {
                output.putAll(error);
            } else if (!createHeader(config.getString(Constants.PARTICIPANT_CODE), recipientCode, apiCallId, correlationId, workflowId, actionJwe, onActionStatus, headers, error)) {
                output.putAll(error);
            } else if (!encryptPayload(headers, fhirPayload, output, config)) {
                output.putAll(error);
            } else {
                result = initializeHCXCall(JSONUtils.serialize(output), operation, response, config);
                output.putAll(response);
            }
            if (output.containsKey(Constants.ERROR) || output.containsKey(ErrorCodes.ERR_DOMAIN_PROCESSING.toString()))
                logger.error("Error while processing the outgoing request: {}", output);
            return result;
        } catch (Exception ex) {
            // TODO: Exception is handled as domain processing error, we will be enhancing in next version.
            ex.printStackTrace();
            output.put(ErrorCodes.ERR_DOMAIN_PROCESSING.toString(), ex.getMessage());
            logger.error("Error while processing the outgoing request: {}", ex.getMessage());
            return result;
        }
    }

    public boolean validatePayload(String fhirPayload, Operations operation, Map<String,Object> error, Config config) {
        if (config.getBoolean(Constants.FHIR_VALIDATION_ENABLED))
            return validateFHIR(fhirPayload, operation, error, config);
        else return true;
    }

    @Override
    public boolean createHeader(String senderCode, String recipientCode, String apiCallId, String correlationId, String workflowId, String actionJwe, String onActionStatus, Map<String, Object> headers, Map<String, Object> error) {
        try {
            headers.put(Constants.ALG, Constants.A256GCM);
            headers.put(Constants.ENC, Constants.RSA_OAEP);
            headers.put(Constants.HCX_TIMESTAMP, new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").format(new Date()));
            if (StringUtils.isEmpty(apiCallId))
                apiCallId = UUID.randomUUID().toString();
            headers.put(Constants.HCX_API_CALL_ID, apiCallId);
            logger.info("recipient key: {}", recipientCode);
            if (!StringUtils.isEmpty(recipientCode)) {
                headers.put(Constants.HCX_SENDER_CODE, senderCode);
                headers.put(Constants.HCX_RECIPIENT_CODE, recipientCode);
                if (StringUtils.isEmpty(correlationId))
                    correlationId = UUID.randomUUID().toString();
                headers.put(Constants.HCX_CORRELATION_ID, correlationId);
                if (!StringUtils.isEmpty(workflowId))
                    headers.put(Constants.WORKFLOW_ID, workflowId);
            } else {
                Map<String, Object> actionHeaders = JSONUtils.decodeBase64String(actionJwe.split("\\.")[0], Map.class);
                headers.put(Constants.HCX_SENDER_CODE, actionHeaders.get(Constants.HCX_RECIPIENT_CODE));
                headers.put(Constants.HCX_RECIPIENT_CODE, actionHeaders.get(Constants.HCX_SENDER_CODE));
                headers.put(Constants.HCX_CORRELATION_ID, actionHeaders.get(Constants.HCX_CORRELATION_ID));
                headers.put(Constants.STATUS, onActionStatus);
                if (actionHeaders.containsKey(Constants.WORKFLOW_ID))
                    headers.put(Constants.WORKFLOW_ID, actionHeaders.get(Constants.WORKFLOW_ID));
            }
            logger.info("Request headers are created: " + headers);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            error.put(Constants.ERROR, "Error while creating headers: " + e.getMessage());
            return false;
        }
    }

    @Override
    public boolean encryptPayload(Map<String, Object> headers, String fhirPayload, Map<String, Object> output, Config config) throws Exception {
        try {
            String authToken = Utils.generateToken(config);
            String publicKeyUrl = (String) Utils.searchRegistry(headers.get(Constants.HCX_RECIPIENT_CODE).toString(), authToken, config.getString(Constants.PROTOCOL_BASE_PATH)).get(Constants.ENCRYPTION_CERT);
            String certificate = IOUtils.toString(new URL(publicKeyUrl), StandardCharsets.UTF_8.toString());
            InputStream stream = new ByteArrayInputStream(certificate.getBytes());
            Reader fileReader = new InputStreamReader(stream);
            RSAPublicKey rsaPublicKey = PublicKeyLoader.loadPublicKeyFromX509Certificate(fileReader);
            JweRequest jweRequest = new JweRequest(headers, JSONUtils.deserialize(fhirPayload, Map.class));
            jweRequest.encryptRequest(rsaPublicKey);
            output.putAll(jweRequest.getEncryptedObject());
            logger.info("Payload is encrypted successfully");
            return true;
        } catch (Exception e) {
            logger.error("Error while encrypting the payload: {}", e.getMessage());
            e.printStackTrace();
            output.put(Constants.ERROR,"Error while encrypting the payload: "+ e.getMessage());
            return false;
        }
    }

    // Exception is handled in processFunction method
    @Override
    public boolean initializeHCXCall(String jwePayload, Operations operation, Map<String, Object> response, Config config) throws Exception {
        return Utils.initializeHCXCall(jwePayload, operation, response, config);
    }

    @Override
    public boolean sendNotification(String topicCode, String recipientType, List<String> recipients, String message, Map<String, String> templateParams, String correlationId, Map<String, Object> output, Config config) throws Exception {
        try {
            NotificationRequest notificationRequest = new NotificationRequest(topicCode, message, templateParams, recipientType, recipients, correlationId, config);
            NotificationService.validateNotificationRequest(notificationRequest);
            Map<String, Object> requestBody = NotificationService.createNotificationRequest(notificationRequest, output, message);
            return initializeHCXCall(JSONUtils.serialize(requestBody), Operations.NOTIFICATION_NOTIFY, output, config);
        } catch (Exception e) {
            logger.error("Error while sending the notification: {}", e.getMessage());
            output.put(Constants.ERROR, "Error while sending the notifications: " + e.getMessage());
            return false;
        }
    }
}
